### PR TITLE
Update report on events

### LINKED ISSUE #144: Add topic listener process
**Describe the new feature you would like to see**
A topic listener process (insert better name here) should be added in order to listen to communication from ISAR / the robot over MQTT or through API calls if necessary to update the current status of executing missions, ISAR state and robot availability. 

This process should listen to relevant topics and update objects (reports and robots) in the database when new information is received.

**Describe the solution you'd like**
Although there is no required solution it is most likely another thread being spun up at startup in `main`. Which listens to predefined topics. Keep in mind that events on topics will lead to this thread modifying the database so thread safe thinking should be important.

**How will this feature affect the current Threat Model?**
Not affected.

### LINKED ISSUE #212: Add database models for Task and Step to reflect ISAR/Echo
**Describe the improvement you would like to see**
ISAR uses Tasks and Steps to represent the structure of the mission, similar to the structure from Echo. This structure should be used in Flotilla as well, were each Report contains a list of Tasks and each Task contain several steps. The database models and API models should be updated to support this structure. Relevant data (time, status, files etc) should be attached to the Task and Step.

**How will this change existing functionality?**
Database will better represent the structure in connected services.

**How will this improvement affect the current Threat Model?**
No change

### PR DESCRIPTION
**Pull request by aestene**
Closes #212
Closes #144

### CODE REVIEW

File: backend/api/Database/Models/IsarStep.cs [hunk 1]
```diff
@@ -1,39 +1,69 @@
...
+    public enum IsarStepStatus
+    {
+        Successful,
+        InProgress,
+        NotStarted,
+        Failed,
+        Cancelled
+    }
+
+    public enum StepType
```
Comment (aestene) [inline backend/api/Database/Models/IsarStep.cs:46]: Not very happy about the "duplication" of `StepType` and `InspectionType` here but not certain what else to do.

**Comment by aestene** [comment]
There is one issue which I'm uncertain how should be addressed.

When starting a mission in ISAR from Flotilla the response from ISAR contains the mission details which Flotilla then writes to the database as a `Report` object which contains `IsarTask` and `IsarStep` as lists. However, the state machine in ISAR has started executing the current mission before Flotilla has written to the database. Thus ISAR is sending status updates for mission, task and step which Flotilla will not recognize because it can't find those IDs in the database yet as the write operation is not complete. 

Not entirely certain how this should be resolved.

**Comment by aestene** [comment]
There is one issue which I'm uncertain how should be addressed.

When starting a mission in ISAR from Flotilla the response from ISAR contains the mission details which Flotilla then writes to the database as a `Report` object which contains `IsarTask` and `IsarStep` as lists. However, the state machine in ISAR has started executing the current mission before Flotilla has written to the database. Thus ISAR is sending status updates for mission, task and step which Flotilla will not recognize because it can't find those IDs in the database yet as the write operation is not complete. 

Not entirely certain how this should be resolved.

**Comment by oysand** [comment]
I think a better way to solve this issue is to move the misssion planner from ISAR to Flotilla so that Flotilla provides the mission to ISAR. That way we can toggle if ISAR should use a local mission planner or be used in combination with Flotilla. We can then write the mission to the database before sending it to ISAR

**Comment by oysand** [comment]
> There is one issue which I'm uncertain how should be addressed.
> 
> When starting a mission in ISAR from Flotilla the response from ISAR contains the mission details which Flotilla then writes to the database as a `Report` object which contains `IsarTask` and `IsarStep` as lists. However, the state machine in ISAR has started executing the current mission before Flotilla has written to the database. Thus ISAR is sending status updates for mission, task and step which Flotilla will not recognize because it can't find those IDs in the database yet as the write operation is not complete.
> 
> Not entirely certain how this should be resolved.

I think a better way to solve this issue is to move the misssion planner from ISAR to Flotilla so that Flotilla provides the mission to ISAR. That way we can toggle if ISAR should use a local mission planner or be used in combination with Flotilla. We can then write the mission to the database before sending it to ISAR

**Comment by aestene** [comment]
> > There is one issue which I'm uncertain how should be addressed.
> > When starting a mission in ISAR from Flotilla the response from ISAR contains the mission details which Flotilla then writes to the database as a `Report` object which contains `IsarTask` and `IsarStep` as lists. However, the state machine in ISAR has started executing the current mission before Flotilla has written to the database. Thus ISAR is sending status updates for mission, task and step which Flotilla will not recognize because it can't find those IDs in the database yet as the write operation is not complete.
> > Not entirely certain how this should be resolved.
> 
> I think a better way to solve this issue is to move the misssion planner from ISAR to Flotilla so that Flotilla provides the mission to ISAR. That way we can toggle if ISAR should use a local mission planner or be used in combination with Flotilla. We can then write the mission to the database before sending it to ISAR

I have mentioned this in a new issue #226. It's easier to solve in a different PR and to just include this functionality as it is now. It works somewhat for turtlebot, but there are some unfortunate sideffects.

**Review by GodVenn — COMMENTED** [review]
LGTM, just one renaming suggestion

File: backend/api/Database/Models/IsarStep.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        Successful,
+        InProgress,
+        NotStarted,
+        Failed,
+        Cancelled
+    }
+
+    public static class IsarStepStatusMethods
+    {
+        public static IsarStepStatus From(string status)
```
Comment (GodVenn) [inline backend/api/Database/Models/IsarStep.cs:48]:
```suggestion
        public static IsarStepStatus FromString(string status)
```

File: backend/api/Database/Models/IsarStep.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        DriveToPose,
+        TakeImage,
+        TakeThermalImage,
+        Audio
+    }
+
+    public class SelectStepType
+    {
+        public static StepType From(string sensorType)
```
Comment (GodVenn) [inline backend/api/Database/Models/IsarStep.cs:80]:
```suggestion
        public static StepType FromSensorTypeAsString(string sensorType)
```

File: backend/api/Database/Models/IsarStep.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    public enum InspectionType
+    {
+        Image,
+        ThermalImage,
+        Audio
+    }
+
+    public class SelectInspectionType
+    {
+        public static InspectionType From(string sensorType)
```
Comment (GodVenn) [inline backend/api/Database/Models/IsarStep.cs:102]:
```suggestion
        public static InspectionType FromSensorTypeAsString(string sensorType)
```

**Review by GodVenn — COMMENTED** [review]
Some naming suggestions for methods

**Review by oysand — APPROVED** [review]
LGTM

File: backend/api/Database/Models/IsarStep.cs (hunk 2 above)
Comment (aestene) [reply backend/api/Database/Models/IsarStep.cs:48]: Done

File: backend/api/Database/Models/IsarStep.cs (hunk 4 above)
Comment (aestene) [reply backend/api/Database/Models/IsarStep.cs:102]: Done

File: backend/api/Database/Models/IsarStep.cs (hunk 3 above)
Comment (aestene) [reply backend/api/Database/Models/IsarStep.cs:80]: Done

File: backend/api/Database/Models/IsarStep.cs [hunk 5]
```diff
@@ -4,36 +4,66 @@
...
+        Failed,
+        Cancelled
+    }
+
+    public enum StepType
+    {
+        DriveToPose,
+        TakeImage,
+        TakeThermalImage,
+        Audio
```
[the remark this answers is not in the document]
Comment (aestene) [reply backend/api/Database/Models/IsarStep.cs:75]: Done

**Review by GodVenn — APPROVED** [review]
LGTM

**Pull request merged**